openmsupply/application-manager-web-app
# #488 "Other" option for radio button selector plugin

Implements #488. Requires [back-end PR branch #289](https://github.com/openmsupply/application-manager-server/pull/289) to see it in action.

~~_Branched from #498 -- please review that one first_~~  Merged

---

Part of [CreateCompany](https://github.com/openmsupply/application-manager-web-app/issues/96) and [JoinCompany](https://github.com/openmsupply/application-manager-web-app/issues/475) epics

To get this working smoothly, I re-factored similarly to the Drop-down changes in #498 so that it uses the "index" as the primary local "value", rather than the passed in "value" which is just the `text` field. It seems to be much easier for most element plugins to handle their "value" locally like this, so the `value/setValue` parameters passed down aren't terribly useful for most plugins.

Also like Drop-down, it can also handle a whole-object array for the options list, as long as a `optionsDisplayProperty` is specified (see #498). 

Anyway, added "hasOther" and "otherPlaceholder" parameters (see docs), and changed the saved Response object as follows (also in docs):

```
{
  optionIndex: <integer> (index from the options array)
  text: <string> (actual text from options array, will be "Other: <value>" if "Other" is selected)
  selection: <string | object> (entire object or string from the supplied options list)
  other: <string> (text of the "Other" selection, only present if "Other" is selected)
}

```

### To test:
Load a Feature Showcase application (using the above mentioned branch) and navigate to Section 1 -> Page 3. The first element on the page should have the "Other" option. 

### Notes

- It would look a lot better if the "Other" text field was inline with the "Other" button label. However, this it not trivial to do, since both `<Radio>` and `<Input>` are block-level elements, and the Radios form a repeating list. Something to think about...
- Let me know if you think the saved Response shape is okay. Another possibility is to not have the "other" field and just set "selection" to the Other value. But then you don't have a record of whether or not "Other" was selected.
- Obviously it'd be advisable to only provide an array of strings as options if you're going to use the `hasOther` parameter, otherwise you'd be mixing-and-matching objects with strings (the "Other" option is always just a string). I haven't tested what happens if you try, but I imagine the results are not pretty.
- Have not yet tested Validation for the "Other" text field. In theory it should work fine, but I'll check it properly when I update the actual "createCompany" template.
